### PR TITLE
V2A AAD Changes 

### PR DESCRIPTION
**Pull request by viverm**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1 [hunk 1]
```diff
@@ -12,7 +12,7 @@
 # RootModule = ''
 
 # Version number of this module.
-ModuleVersion = '0.1.6'
+ModuleVersion = '0.1.5'
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1]: @viverm looks like a bad merge caused some unnecessary changes in this file

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery.Test/ScenarioTests/AsrTestsBase.cs [hunk 2]
```diff
@@ -31,33 +31,57 @@
...
                 try
-                {
-                    var serializer1 = new DataContractSerializer(typeof(ASRVaultCreds));
-                    using (var s = new FileStream(
-                        this.vaultSettingsFilePath,
-                        FileMode.Open,
-                        FileAccess.Read,
-                        FileShare.Read))
+                {  
+                    if (File.ReadAllText(this.vaultSettingsFilePath).ToLower().Contains("<asrvaultcreds"))
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery.Test/ScenarioTests/AsrTestsBase.cs]: @viverm to read from/write to files, use `DataStore.ReadFromFileStream()` from in the `FileUtilities` class in `Commands.Common`

https://github.com/Azure/azure-powershell/blob/preview/src/Common/Commands.Common/Utilities/FileUtilities.cs#L41

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+using System.Management.Automation;
+using System.Threading;
+using Microsoft.Azure.Management.RecoveryServices.SiteRecovery.Models;
+
+namespace Microsoft.Azure.Commands.RecoveryServices.SiteRecovery
+{
+    /// <summary>
+    ///     Sets Azure Site Revcovery Notification / Alert.
+    /// </summary>
+    [Cmdlet(
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs]: @viverm it looks like you're missing the `SupportsShouldProcess` property in this cmdlt attribute

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        "Set-AzureRmRecoveryServicesAsrNotificationSetting",
+        "Set-ASRAlertSetting")]
+    [OutputType(typeof(ASRAlertSetting))]
+    public class SetAzureRmRecoveryServicesAsrAlertSetting : SiteRecoveryCmdletBase
+    {
+        /// <summary>
+        ///     Gets or sets EmailSubscriptionOwner .Mail to subscription owner.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.EmailToSubscriptionOwner, Mandatory = true)]
+        public SwitchParameter EmailSubscriptionOwners { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs]: @viverm this should be renamed to `EnableEmailSubscriptionOwners`

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        /// </summary>
+        public override void ExecuteSiteRecoveryCmdlet()
+        {
+            base.ExecuteSiteRecoveryCmdlet();
+            if (this.ShouldProcess(
+                "Alerts",
+                VerbsCommon.Set))
+            {
+                base.ExecuteSiteRecoveryCmdlet();
+                if (this.ShouldProcess(
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs]: @viverm why is there two calls to `this.ShouldProcess` and `base.ExecuteSiteRecoveryCmdlet`?

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    public class GetAzureRmRecoveryServicesAsrEvents : SiteRecoveryCmdletBase
+    {
+        /// <summary>
+        ///     Gets or sets Resource Id.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.ByResourceId, Mandatory = true)]
+        public string ResourceId { get; set; }
+
+        /// <summary>
+        ///     Gets or sets Resource Id.
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: @viverm typo in this description?

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        [ValidateNotNullOrEmpty]
+        public DateTime EndTime { get; set; }
+
+        /// <summary>
+        ///     Gets or sets the fabric.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.ByParam)]
+        [Parameter(ParameterSetName = ASRParameterSets.ByFabricId)]
+        [ValidateNotNullOrEmpty]
+        public ASRFabric Fabric { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: @viverm does `Fabric` contain a `FabricId`?

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        ///     Gets or sets Resource Id.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.ByResourceId, Mandatory = true)]
+        public string ResourceId { get; set; }
+
+        /// <summary>
+        ///     Gets or sets Resource Id.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.ByFabricId, Mandatory = true)]
+        public string FabricId { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: @viverm what exactly is `FabricId`?

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            ASRHealthEventServerity.Unknown)]
+        public string Severity { get; set; }
+
+        /// <summary>
+        ///     Gets or sets the Start time.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.ByParam)]
+        [Parameter(ParameterSetName = ASRParameterSets.ByFabricId)]
+        [ValidateNotNullOrEmpty]
+        public DateTime StartTime { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: @viverm should there be a corresponding `EndTime` parameter?

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/ReplicationProtectedItem/StartAzureRmRecoveryServicesAsrResynchronizeReplicationJob.cs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        "Start-ASRResynchronizeReplicationJob",
+        "Start-ASRResyncJob")]
+    [OutputType(typeof(ASRJob))]
+    public class StartAzureRmRecoveryServicesAsrResynchronizeReplicationJob : SiteRecoveryCmdletBase
+    {
+        /// <summary>
+        ///     Gets or sets Resource Id.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.ByResourceId, Mandatory = true)]
+        public string ResourceId { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/ReplicationProtectedItem/StartAzureRmRecoveryServicesAsrResynchronizeReplicationJob.cs]: @viverm anywhere that you are using a `ResourceId` parameter, it should have the property `ValueFromPipelineByPropertyName = true` so users are able to pipe the result of `Get/Find-AzureRmResource` to a cmdlet with a `ResourceId` parameter

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/NewAzureRmRecoveryServicesAsrVCenter.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    [Alias("New-ASRvCenter")]
+    [OutputType(typeof(ASRJob))]
+    public class NewAzureRmRecoveryServicesAsrvCenter : SiteRecoveryCmdletBase
+    {
+        /// <summary>
+        ///     Gets or sets Fabric of the vCenter.
+        /// </summary>
+        [Parameter(Mandatory = true, ValueFromPipeline = true)]
+        [ValidateNotNullOrEmpty]
+        public ASRFabric Fabric { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/NewAzureRmRecoveryServicesAsrVCenter.cs]: @viverm parameter set for `ResourceId` of Fabric being provided? Please make sure that all cmldets that should support piping have the following three scenarios enabled:

1) By object - involves using an `InputObject` parameter that has `ValueFromPipeline = true`
2) By resource id - involves using a `ResourceId` parameter that has `ValueFromPipelineByPropertyName = true`
3) By command line - involves using a `Name` and `ResourceGroupName` parameter with no ability to accept its value from the pipeline; this is your default parameter set

File: src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1 (hunk 1 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1]: Changed to 0.1.6

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery.Test/ScenarioTests/AsrTestsBase.cs (hunk 2 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery.Test/ScenarioTests/AsrTestsBase.cs]: Used FileUtilities.DataStore.ReadFileAsText method

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs (hunk 3 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs]: That was the miss . Added that again.

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs (hunk 4 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs]: Changed

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs (hunk 5 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/AlertSetting/SetAzureRmRecoveryServicesAsrAlertSetting.cs]: removed

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs (hunk 6 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: corrected

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs (hunk 8 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: Id of ASRFabric.

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs (hunk 7 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: yes

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs (hunk 9 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: Already there . But reordered below StartTime to give better visibility.

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/ReplicationProtectedItem/StartAzureRmRecoveryServicesAsrResynchronizeReplicationJob.cs (hunk 10 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/ReplicationProtectedItem/StartAzureRmRecoveryServicesAsrResynchronizeReplicationJob.cs]: Done

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/NewAzureRmRecoveryServicesAsrVCenter.cs (hunk 11 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/NewAzureRmRecoveryServicesAsrVCenter.cs]: First two scenario fits for ASR. 
Getting the PSObject by Name and ResourceGroupName parameter is not possible in ASR Case we need additional parameter like vault , protection container etc on case to case basis.
Name is mainly server generated Id and hard to remember.

**Comment by markcowl** [comment]
@viverm It looks like you haven't committed and pushed the changes you mention in your responses to review comments.

**Comment by viverm** [comment]
Error during build 

[https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/322/console](https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/322/console )

Saw similar error with other PR builds/

d:\workspace\powershell\src\ResourceManager\Common\Commands.ScenarioTests.ResourceManager.Common\EnvironmentSetupHelper.cs(427): error : Microsoft.Azure.Commands.Sql.Test.ScenarioTests.VirtualNetworkRuleTest.TestVirtualNetworkRuleRemove [FAIL] [d:\workspace\powershell\build.proj]
04:47:22   d:\workspace\powershell\src\ResourceManager\Common\Commands.ScenarioTests.ResourceManager.Common\EnvironmentSetupHelper.cs(427): error : Microsoft.Azure.Commands.Sql.Test.ScenarioTests.VirtualNetworkRuleTest.TestVirtualNetworkRuleCreateAndUpdate [FAIL] [d:\workspace\powershell\build.proj]
04:47:22   d:\workspace\powershell\src\ResourceManager\Common\Commands.ScenarioTests.ResourceManager.Common\EnvironmentSetupHelper.cs(427): error : Microsoft.Azure.Commands.Sql.Test.ScenarioTests.VirtualNetworkRuleTest.TestVirtualNetworkRuleGet [FAIL] [d:\workspace\powershell\build.proj]

File: src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1 [hunk 12]
```diff
@@ -170,7 +231,7 @@ PrivateData = @{
         Tags = 'Azure','ResourceManager','ARM','SiteRecovery'
 
         # A URL to the license for this module.
-        LicenseUri = 'https://aka.ms/azps-license'
+        LicenseUri = 'https://raw.githubusercontent.com/Azure/azure-powershell/preview/LICENSE.txt'
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1]: @viverm please revert this change

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Vault/SetAzureRmRecoveryServicesAsrVaultSettings.cs [hunk 13]
```diff
@@ -75,25 +76,28 @@ public override void ExecuteSiteRecoveryCmdlet()
             {
                 using (var powerShell = System.Management.Automation.PowerShell.Create())
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Vault/SetAzureRmRecoveryServicesAsrVaultSettings.cs]: @viverm if the `powerShell` object is not being used anymore, we should remove the using

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Utilities/Utilities.cs [hunk 14]
```diff
@@ -153,7 +155,7 @@ public static string GetDefaultPath()
                 throw new Exception("Invalid ARM ID");
             }
 
-            for (var i = 1; i < armFields.Length; i = i + 2)
+            for (var i = 0; i < armFields.Length; i = i + 2)
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Utilities/Utilities.cs]: @viverm shouldn't this _always_ fail with an `IndexOutOfRangeException`?

For example, if `armFields` has five elements, this loop will get the following elements:

1st (idx 0) and 2nd (idx 1)
3rd (idx 2

**Comment by cormacpayne** [comment]
@viverm https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/117/

File: src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1 (hunk 12 above)
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1]: @viverm please revert this change

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Utilities/Utilities.cs (hunk 14 above)
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Utilities/Utilities.cs]: @viverm shouldn't this _always_ fail with an `IndexOutOfRangeException`?

For example, if `armFields` has five elements, this loop will get the following elements:

`i == 0`: 1st (idx 0) and 2nd (idx 1)
`i == 2`: 3rd (idx 2) and 4th (idx 3)
`i == 4`: 5th (idx 4) and 6th (idx 5 **DOESN'T EXIST!**)

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Vault/SetAzureRmRecoveryServicesAsrVaultSettings.cs (hunk 13 above)
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Vault/SetAzureRmRecoveryServicesAsrVaultSettings.cs]: @viverm if the `powerShell` object is not being used anymore, we should remove the using

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/RecoveryPlan/UpdateAzureRmRecoveryServicesAsrRecoveryPlan.cs [hunk 15]
```diff
@@ -148,6 +148,20 @@ public override void ExecuteSiteRecoveryCmdlet()
                                             VmId = ((HyperVReplicaBlueReplicationDetails)item
                                                 .Properties.ProviderSpecificDetails).VmId;
                                         }
+                                        else if (item.Properties.ProviderSpecificDetails
+                                                .GetType() ==
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/RecoveryPlan/UpdateAzureRmRecoveryServicesAsrRecoveryPlan.cs]: @viverm nit: for the lines where `.GetType()` is by itself, can you tab it to be consistent with the other instances of it above

It looks like this is also the case in `NewAzureRmRecoveryServicesAsrRecoveryPlan.cs`

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Utilities/Utilities.cs (hunk 14 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Utilities/Utilities.cs]: here we were expecting things in pair  like 1st will be the 2nd the value .
We are using the other utility method to get the  value .this call is not used anymore .
Marked this as part of cleanup activity,

File: src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1 (hunk 12 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/AzureRM.RecoveryServices.SiteRecovery.psd1]: ok

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/RecoveryPlan/UpdateAzureRmRecoveryServicesAsrRecoveryPlan.cs (hunk 15 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/RecoveryPlan/UpdateAzureRmRecoveryServicesAsrRecoveryPlan.cs]: ok

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+        [ValidateNotNullOrEmpty]
+        public string AffectedObjectFriendlyName { get; set; }
+
+        /// <summary>
+        ///     Gets or sets the fabric.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.ByParam)]
+        [Parameter(ParameterSetName = ASRParameterSets.ByFabricId)]
+        [ValidateNotNullOrEmpty]
+        public ASRFabric Fabric { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: @viverm raising this again - if a `Fabric` has a `FabricId`, why does the user need to provide one? It seems like you're just using it to get the `Name` of the fabric, but you already receive the name from the mandatory `Fabric` parameter that a user would pass.

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/RemoveAzureRmRecoveryServicesAsrVCenter.cs [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+            Mandatory = true,
+            ValueFromPipeline = true)]
+        [ValidateNotNullOrEmpty]
+        [Alias("vCenter")]
+        public ASRvCenter InputObject { get; set; }
+
+        /// <summary>
+        ///     ProcessRecord of the command.
+        /// </summary>
+        public override void ExecuteSiteRecoveryCmdlet()
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/RemoveAzureRmRecoveryServicesAsrVCenter.cs]: @viverm can we add a parameter set that allows users to provide a `Fabric` and `Name`? This will be the default parameter set that will take care of the command line scenario and is similar to what is done with the `Get-AsrVCenter` cmdlet.

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/RemoveAzureRmRecoveryServicesAsrVCenter.cs [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+                case ASRParameterSets.Default:
+                    vCenterName = this.InputObject.Name;
+                    fabricName = this.InputObject.FabricArmResourceName;
+                    break;
+            }
+
+            if (this.ShouldProcess(vCenterName, VerbsCommon.Remove))
+            {
+                var response = this.RecoveryServicesClient.RemoveAzureRmSiteRecoveryvCenter(
+                    fabricName,
```
Comment (cormacpayne) [inline src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/RemoveAzureRmRecoveryServicesAsrVCenter.cs]: @viverm just a general question for cmdlets that have a `Fabric` parameter (and this might have been discussed already): why don't we make the user provide a `FabricName` parameter instead of the entire object? That way, they don't have to get the `Fabric` object each time they want to run these cmdlets and can just provide a string instead. Is the `FabricName` hard to discover for the user, or difficult to remember?

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs (hunk 16 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/Event/GetAzureRmRecoveryServicesAsrEvent.cs]: We can make is mutually exclusive and break in two paramset. If the user has fabricId or fabricObject he can directly call it.

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/RemoveAzureRmRecoveryServicesAsrVCenter.cs (hunk 17 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/RemoveAzureRmRecoveryServicesAsrVCenter.cs]: sure.

File: src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/RemoveAzureRmRecoveryServicesAsrVCenter.cs (hunk 18 above)
Comment (viverm) [reply src/ResourceManager/RecoveryServices.SiteRecovery/Commands.RecoveryServices.SiteRecovery/VCenter/RemoveAzureRmRecoveryServicesAsrVCenter.cs]: Fabric Name  is ASR generated entity which is unique .  It is Difficult for user to remember and pass it. user will store and pass it so he can fabric object like he is doing in other cmdlets.

**Comment by cormacpayne** [comment]
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/140/

**Comment by viverm** [comment]
@cormacpayne   Build Completed 
https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/140/console
